src/main/java/org/cloudfoundry/promregator/lite/cfaccessor/SimpleCFAccessorImpl.kt

```diff
@@ -0,0 +1,511 @@
+        }
+//        val key = String.format("%s|%s", orgId, spaceId)
+//        val requestGenerator = PaginatedRequestGeneratorFunctionV3 { resultsPerPage: Int, pageNumber: Int ->
+//
+//        }
+//        val responseGenerator = PaginatedResponseGeneratorFunctionV3 { list: List<org.cloudfoundry.client.v3.applications.ApplicationResource?>, numberOfPages: Int ->
+//            org.cloudfoundry.client.v3.applications.ListApplicationsResponse.builder()
+//                    .addAllResources(list)
+//                    .pagination(Pagination.builder().totalPages(numberOfPages).totalResults(list.size).build())
+//                    .build()
+//        }
+//        return paginatedRequestFetcher.performGenericPagedRetrievalV3(RequestType.ALL_APPS_IN_SPACE, key, requestGenerator,
+//                { r: org.cloudfoundry.client.v3.applications.ListApplicationsRequest? ->
+//                    cloudFoundryClients[api]?.applicationsV3()?.list(r) ?: Mono.empty() },
+//                cf.request.timeout.appInSpace, responseGenerator)
+
+        return cfQueryV3<ApplicationResourceV3, ListApplicationsResponseV3, ListApplicationsRequestV3>(api, ALL_APPS_IN_SPACE) {
+            request { pageNumber ->
+                ListApplicationsRequestV3.builder()
+                        .organizationId(orgId)
+                        .spaceId(spaceId)
+                        .perPage(100)
+                        .page(pageNumber)
+                        .build()
+            }
+            query { applicationsV3().list(it) }
+        }
+    }
+
+    private suspend fun <R: ResourceV2<*>, T: org.cloudfoundry.client.v2.PaginatedResponse<R>,S : PaginatedRequestV2> cfQueryV2(api: String, type: RequestType, requestBlock: CfClientDslV2<R, T, S>.() -> Unit): List<R> {
+        val cfClient = cloudFoundryClients[api]
+        val dsl = CfClientDslV2<R, T, S>(cfClient, type)
+        dsl.requestBlock()
+        return dsl.runQueries()
+    }
+
+    private suspend fun <R: Resource, T: org.cloudfoundry.client.v3.PaginatedResponse<R>,S : V3PaginatedRequest> cfQueryV3(api: String, type: RequestType, requestBlock: CfClientDslV3<R, T, S>.() -> Unit): List<R> {
+        val cfClient = cloudFoundryClients[api]
+        val dsl = CfClientDslV3<R, T, S>(cfClient, type)
+        dsl.requestBlock()
+        return dsl.runQueries()
+    }
+
+    inner class CfClientDslV2<R: ResourceV2<*>, T: org.cloudfoundry.client.v2.PaginatedResponse<R>, S: PaginatedRequestV2>(
+            private val cfClient: ReactorCloudFoundryClient?,
+            private val type: RequestType,
+            ) {
+        lateinit var requestTemplate : (Int) -> S
+        lateinit var queryTemplate: ReactorCloudFoundryClient.(S) -> Mono<T>
+
+        fun request(block: (Int) -> S) {
+            requestTemplate = block
+        }
+
+        fun query(block: ReactorCloudFoundryClient.(S) -> Mono<T>) {
+            queryTemplate = block
+        }
+
+        suspend fun runQueries() : List<R> {
```

Comment:
This method is where the brains of the query is.

It makes the `first` call to get the total number of pages (wrapping it in a ratelimit and timer) then fetches all the remaining pages in parallel. (See line 425)